awx_collection/plugins/modules/tower_credential_input_source.py

```diff
@@ -0,0 +1,134 @@
+
+
+DOCUMENTATION = '''
+---
+module: tower_credential_input_source
+author: "Tom Page (@Tompage1994)"
+version_added: "2.3"
+short_description: create, update, or destroy Ansible Tower credential input sources.
+description:
+    - Create, update, or destroy Ansible Tower credential input sources. See
+      U(https://www.ansible.com/tower) for an overview.
+options:
+    description:
+      description:
+        - The description to use for the credential input source.
+      type: str
+    input_field_name:
+      description:
+        - The input field the credential source will be used for
+      required: True
+      type: str
+    metadata:
+      description:
+        - A JSON or YAML string
+      required: False
+      type: str
+    target_credential:
+      description:
+        - The credential which will have its input defined by this source
+      required: true
+      type: str
+    source_credential:
+      description:
+        - The credential which is the source of the credential lookup
+      required: true
+      type: str
+    state:
+      description:
+        - Desired state of the resource.
+      choices: ["present", "absent"]
+      default: "present"
+      type: str
+
+extends_documentation_fragment: awx.awx.auth
+'''
+
+
+EXAMPLES = '''
+- name: Use CyberArk Lookup credential as password source
+  tower_credential_input_source:
+    input_field_name: password
+    target_credential: new_cred
+    source_credential: cyberark_lookup
+    metadata:
+      object_query: "Safe=MY_SAFE;Object=awxuser"
+      object_query_format: "Exact"
+    state: present
+
+'''
+
+from ..module_utils.tower_api import TowerModule
+
+def main():
+    # Any additional arguments that are not fields of the item can be added here
+    argument_spec = dict(
+        description=dict(default=''),
+        input_field_name=dict(required=True),
+        target_credential=dict(required=True),
+        source_credential=dict(required=True),
+        metadata=dict(type=dict),
+        state=dict(choices=['present', 'absent'], default='present'),
+    )
+
+    # Create a module for ourselves
+    module = TowerModule(argument_spec=argument_spec)
+
+    # Extract our parameters
+    description = module.params.get('description')
+    input_field_name = module.params.get('input_field_name')
+    target_credential = module.params.get('target_credential')
+    source_credential = module.params.get('source_credential')
+    metadata = module.params.get('metadata')
+    state = module.params.get('state')
+
+    target_credential_id = module.resolve_name_to_id('credentials', target_credential)
+    source_credential_id = module.resolve_name_to_id('credentials', source_credential)
+
+    # Attempt to look up the object based on the provided name, credential type and optional organization
+    lookup_data = {
+        'target_credential': target_credential_id,
+        'source_credential': source_credential_id,
+        'input_field_name': input_field_name,
```

Comment:
It looks like the uniqueness rule is

```python
        unique_together = (('target_credential', 'input_field_name'),)
```

So the lookup by the module should probably correspond, and this only affects `source_credential`.

Let me talk that out. With the API, you cannot create 2 credentials, both with the same `target_credential` and `input_field_name`, but different `source_credential`s, because this would violate the uniqueness rule.

Logically, I _think_ that makes sense. If 2 source credentials sought to provide the same literal input, then that creates a conflict. @jakemcdermott was involved in this feature development, so I'd like a sanity check from him.

It may even be possible to specify a target credential, field name, and then _change_ the source credential for that field. If the module was given all 3 of these inputs, that is probably what one would expect it to do (if it differed from the existing field).

Reply:
> you cannot create 2 credentials, both with the same target_credential and input_field_name, > but different source_credentials, because this would violate the uniqueness rule.

This is correct although `2 input sources` instead of `2 credentials` would be more accurate.

> It may even be possible to specify a target credential, field name, and then change the source credential for that field. 

Yes, the api should allow this. Ensuring the specified source credential is the actual source credential seems like the idempotent thing to do and so I'm with you on that being the expected behavior.

I'd expect an `input_field` change to work as well provided the new field + credential combination exists and isn't already in use by another input source.

Reply:
This makes sense. I have now updated to this.